packages/core/admin/admin/src/utils/getFetchClient.ts

```diff
@@ -111,20 +103,182 @@ type FetchClient = {
+       * In native Fetch API, a response is considered "not ok"
+       * when the status code falls in the 200 to 299 (inclusive) range
+       */
+      if (!response.ok && result.error && !validateStatus?.(response.status)) {
+        throw new FetchError(result.error.message, { data: result });
+      }
+
+      if (!response.ok && !validateStatus?.(response.status)) {
+        throw new FetchError('Unknown Server Error');
+      }
+
+      return { data: result };
+    } catch (error) {
+      if (error instanceof SyntaxError && response.ok) {
+        // Making sure that a SyntaxError doesn't throw if it's successful
+        return { data: [], status: response.status };
+      } else {
+        throw error;
+      }
+    }
   };
+
+  const paramsSerializer =
+    <Param = unknown>(params?: Param) =>
+    (url: string) => {
+      if (params) {
+        const serializedParams = qs.stringify(params, { encode: false });
+        return `${url}?${serializedParams}`;
+      }
+      return url;
+    };
+
+  const addBaseUrl = (url: FetchURL) => {
+    return `${backendURL}${url}`;
+  };
+
+  /**
+   * We use the factory method because the options
+   * are unique to the individual request
+   */
+  const makeCreateRequestUrl = (options?: FetchOptions) =>
+    pipe(normalizeUrl, addBaseUrl, paramsSerializer(options?.params));
+
+  const fetchClient: FetchClient = {
+    get: async <TData>(url: string, options?: FetchOptions): Promise<FetchResponse<TData>> => {
+      const headers = new Headers({
+        ...defaultHeader,
+        ...options?.headers,
+      });
+      /**
+       * this applies all our transformations to the URL
+       * - normalizing (making sure it has the correct slash)
+       * - appending our BaseURL which comes from the window.strapi object
+       * - serializing our params with QS
+       */
+      const createRequestUrl = makeCreateRequestUrl(options);
+      const response = await fetch(createRequestUrl(url), {
+        signal: options?.signal ?? defaultOptions.signal,
+        method: 'GET',
+        headers,
+      });
+      return responseInterceptor<TData>(response, options?.validateStatus);
+    },
+    post: async <TData, TSend = unknown>(
+      url: string,
+      data?: TSend,
+      options?: FetchOptions
+    ): Promise<FetchResponse<TData>> => {
+      const headers = new Headers({
+        ...defaultHeader,
+        ...options?.headers,
+      });
+
+      const createRequestUrl = makeCreateRequestUrl(options);
+
+      // Todo: remove this and replace it with a native implementation
+      if (isFormDataRequest(headers)) {
+        return axios.post(createRequestUrl(url), data, {
+          headers: {
+            ...defaultHeader,
+            ...options?.headers,
+          },
+          signal: options?.signal ?? defaultOptions.signal,
+          onUploadProgress: options?.onUploadProgress,
+          validateStatus: options?.validateStatus,
+        });
+      }
```

Comment:
Would it be better to have this implementation solely in the upload plugin? that way when we're ready to remove it and sort it out, it's not a breaking change.

Reply:
hmm, well I thought about it the other way around, this makes things a little neater, the upload plugin now has zero axios dependencies, and this is basically just hidden implementation it's shouldn't affect the FetchClient api so changing it later on shouldn't be a breaking change, it's more of making sure that the upload is using the same api as every other package, and axios is only left now in the admin package

Reply:
but what if a plugin does a form request and passes an `onUploadProgress` function? we're then bound to keep this API until V6, we can't remove it because that could break said plugin right?

Reply:
hmmm, yes well my intention was to commit on having the onUploadProgress working as it was,
and have it recreated, in the same way validateStatus or params options have been recreated to keep theses behaviors that we are using in the codebase working.

Reply:
It's a big commitment, considering we'd _most likely_ need to use XMLHttpRequest to support onUploadProgress since the stream API is not in every browser and only works with `http/2+`. We're also not covering all the use cases of axios here e.g. `onDownloadProgress` & I don't think we really should be imho because we're not making an axios clone. I think we're better off keeping this isolated to `upload` only and we can manage it there.